content/200-concepts/100-components/02-prisma-client/052-working-with-generated-types/index.mdx

```diff
@@ -0,0 +1,101 @@
+---
+title: 'Advanced usage of generated types'
+metaTitle: 'Advanced usage of generated TypeScript types (Reference)'
+metaDescription: 'Prisma Client provides full type safety for queries, even for partial queries or included relations. This page explains how to leverage the generated types.'
+---
+
+<TopBlock>
+
+The generated code for Prisma Client contains several helpful types that you can use to make your application more type-safe. This page describes patterns for leveraging some of the generated types.
+
+By importing the `Prisma` namespace from `@prisma/client` you can access types related to your defined models.
+
+The following example composes a `select` param outside of the queries context using the [`validator`](#prisma-validator) function, it passes in a generated type (`UserSelect`) meaning only fields available on a `select` can be used.
+
+```ts
+// Import the `Prisma` namespace from the `@prisma/client` package
+import { PrismaClient, Prisma } from '@prisma/client'
+
+const prisma = new PrismaClient()
+
+// Pass in the UserSelect type assertion
+const selectId = Prisma.validator<Prisma.UserSelect>()({
+  id: true,
+})
+
+const data = await prisma.user.findFirst({ select: selectId })
+```
+
+> **Note**: You can find all type definitions inside the `index.d.ts` file inside the [generated `.prisma/client` folder](working-with-prismaclient/generating-prisma-client#the-prismaclient-npm-package).
+
+</TopBlock>
+
+### <inlinecode>Prisma.validator</inlinecode>
+
+The `validator` is a function that exists on the Prisma namespace. It accepts a generated type and returns an object that fits that types shape. It helps you create re-usable query parameters based on your schema models while making sure that the objects you create are valid.
+
+The following example creates a function from the `validator`, it can then be used when interacting with user created data, such as form inputs.
+
+```ts
+import { Prisma } from '@prisma/client'
+
+const prisma = new PrismaClient()
+
+// Create a new user and a post from user input
+const createUserAndPost = (name: string, email: string, postTitle: string, profileBio: string) => {
+  return Prisma.validator<Prisma.UserCreateInput>()({
+    name,
+    email,
+    posts: {
+      create: {
+        title: postTitle,
+      },
+    },
+    profile: {
+      create: {
+        bio: profileBio,
+      },
+    },
+  })
+}
+
+// Find a specific user from user input
+const findSpecificUser = (name: string, email: string) => {
+  return Prisma.validator<Prisma.UserWhereInput>()({
+    name,
+    email,
+  })
+}
+
+// Create the user in the database
+await prisma.user.create({ data: createUserAndPost })
```

Comment:
when I tried to run this, I got an error in VS Code. is it correct that a _function_ (which is what `createUserAndPost` is) is passed to a `create` query?

Reply:
Oh! Ha! Sorry, i have forgotten to pass in the functions params, thats why its not working.

It should be `await prisma.user.create({ data: createUserAndPost('Rich', 'rich@boop.com', 'Post1', 'Im a writer') })`